Новая версия.

demidko opened the pull request:
Исправлена проблема коллизий хеша. Скорость не просела..

demidko committed:
update

demidko committed:
небольшие косметические правки читабельности

demidko committed:
рефакторинг, улучшения читабельности

Commit:
Create README.md

demidko committed:
Merge remote-tracking branch 'origin/master'

demidko committed:
вынес jmh в отедльный maven модуль

demidko committed:
вынес jmh в отдельный maven модуль

demidko committed:
refactoring

demidko committed:
refactoring

demidko committed:
Исправлена структура проекта.

demidko committed:
Исправлена структура проекта.

demidko committed:
Merge remote-tracking branch 'origin/master'

Commit:
Update README.md

Commit:
Update README.md

demidko committed:
рефакторинг

demidko committed:
Merge remote-tracking branch 'origin/master'

demidko committed:
исправил опечатку

demidko committed:
c

demidko committed:
Добавил описание

Commit:
Update README.md

demidko committed:
Добавил описание

demidko committed:
Еще описание

Commit:
Update README.md

demidko committed:
Добавил модуль компилятора mrd & tab --> bin

demidko committed:
Описание текущих проблем.

demidko committed:
Код разрешает проблему колиззий хеша строк.

demidko committed:
Код разрешает проблему коллизий хеша строк.

demidko committed:
Исправил ошибку - число приводилось к байту и записывалось.
Теперь число корректно преобразуется в 4 байта и записывается.

demidko committed:
Исправил ошибку - число приводилось к байту и записывалось.
Теперь число корректно преобразуется в 4 байта и записывается.

demidko committed:
Добавил корректный бинарный файл, с правильно записанными хешами.

demidko committed:
Добавил корректный бинарный файл, с правильно записанными хешами.

demidko committed:
Исправил ошибку считывания леммы из бинарного файла (из буфера копировался лишний символ).
Добавил тестовое консольное приложение, для "ручного" тестирования.

demidko committed:
Исправил ошибку считывания леммы из бинарного файла (из буфера копировался лишний символ).
Добавил тестовое консольное приложение, для "ручного" тестирования.

demidko committed:
Обновил описание репозитория.

demidko committed:
Обновил описание репозитория.

Commit:
Update README.md

Commit:
Update README.md

demidko committed:
Обновил описание репозитория.

demidko committed:
Обновил описание репозитория.

demidko committed:
Merge remote-tracking branch 'origin/master'

demidko committed:
Исправил описание класса

demidko committed:
Добавил сборку тестового приложения в отдельный jar со всеми зависимостями.

Commit:
Update README.md

Commit:
Удалил неактуальную заметку

demidko committed:
Почистил код компилятора, убрал старые импорты и неиспользуемые файлы.

Commit:
Update README.md

demidko committed:
Упорядочил структуру пакета.

Commit:
Update FlexionStorage.java

demidko committed:
Комментарии

Commit:
Update Decompiler.java

Commit:
Update GrammarInfo.java

Commit:
Update FlexionStorage.java

demidko committed:
Новые тесты

Commit:
Update TestApplication.java

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

demidko committed:
Добавил метод для проверки флексии на грамматическую характеристику. Напр. является ли такая то флексия существительным. Заменил массивы на коллекции, так как в приложении поиска в основном используются коллекции.

demidko committed:
Исправил тесты (вслед за изменениями кода).

bazhenov commented on compiler/src/main/java/com/farpost/aot/mrd/Flexion.java:
```diff
@@ -0,0 +1,21 @@
+package com.farpost.aot.mrd;
+
+public class Flexion {
```
У нас два класса с таким именем. Можем мы их как-то удачней назвать?

demidko replied:
не актуально

bazhenov commented on compiler/src/main/java/com/farpost/aot/mrd/Mod.java:
```diff
@@ -0,0 +1,40 @@
+	public final String postfix, prefix;
+	public final int grammarInfoLineIndex;
+
+	public Mod(final String postfix, final int infoIndex, final String prefix) {
```
Давай писать `final` только там, где неизменность переменной действительно имеет значение (например, на полях класса). В этом конструкторе все аргументы просто копируются в свойства объекта. При этом даже изменение объектов внутри конструктора, не может поменять ссылку снаружи. Повсеместное использование `final` только затрудняет чтение и ничего не добавляет.

demidko replied:
не актуально

bazhenov commented on compiler/src/main/java/com/farpost/aot/mrd/Paradigm.java:
```diff
@@ -0,0 +1,23 @@
+			.collect(Collectors.toList());
+	}
+
+	public Paradigm(final List<Mod> mods) {
```
Конструктор должен быть первый после свойств

demidko replied:
не актуально

bazhenov commented on compiler/src/main/java/com/farpost/aot/Utils.java:
```diff
@@ -0,0 +1,16 @@
+
+final public class Utils {
+
+	public static int hashFromString(final String input) {
```
Чем обусловлено наличие этого метода? Почему ты не используешь `str.hashCode()`?

demidko replied:
не актуально.

bazhenov commented on compiler/src/main/java/com/farpost/aot/Utf8ResourceReader.java:
```diff
@@ -0,0 +1,24 @@
+			));
+	}
+
+	public void close() throws IOException {
```
Если этот класс тоже должен корректно закрываться клиентом, то он должен имплементировать `Closeable`

bazhenov commented on compiler/src/main/java/com/farpost/aot/Main.java:
```diff
@@ -0,0 +1,10 @@
+
+public class Main {
+
+	public static void main(final String[] args) throws IOException {
```
Давай класс `Compilation` и сделаем main-классом

demidko replied:
Ок

bazhenov commented on compiler/src/main/java/com/farpost/aot/Utf8ResourceReader.java:
```diff
@@ -0,0 +1,24 @@
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+
+public abstract class Utf8ResourceReader {
```
Давай избавимся от этого класса, если это возможно. Он не выполняет никакой полезной работы, помимо обертки на `BufferedReader`'ом

demidko committed:
Добавил опциональный метод isNot для флексии.

demidko committed:
Рефакторинг

demidko committed:
Рефакторинг

demidko committed:
Добавил дополнительные, более высокоуровневые методы для класса Flexion.
Исправил описание.

demidko committed:
Исправил описание.

demidko committed:
Привел код в соответствие с принятым стилем.

Commit:
Update FlexionStorage.java

demidko committed:
Исправил документирующие комментарии

demidko committed:
Merge remote-tracking branch 'origin/master'

demidko committed:
План дальнейших изменений.

demidko committed:
План дальнейших изменений.

demidko committed:
Обновил Readme

Commit:
Update README.md

demidko committed:
новый загрузчик mrd словаря

demidko committed:
GrammarInfo refactoring

demidko committed:
new gramtab storage

demidko committed:
new gramtab storage

demidko committed:
Исправление:
Ускоренный парсинг
Вывод предупреждений о неоднозначностях на этапе парсинга
Ускореная компиляция
Ускоренная загрузка
Ускоренный поиск

Commit:
Merge pull request #1 from demidko/compilation-problem

Ошибки исправлены.

demidko committed:
TestApp Java 11 lang level

demidko committed:
Удалил неиспользуемые классы

Commit:
Update README.md

demidko committed:
Добавил тесты.

demidko committed:
deploy settings

demidko committed:
Добавил высоуоуровневые проверки для флексий

bazhenov commented on library/src/main/java/com/farpost/aot/func/Decompiler.java:
```diff
@@ -0,0 +1,52 @@
+		// потому что не является отрицательным (отрицательные зарезрвированы под буквы),
+		// и при этом он больше любого порядкого номера GrammarInfo (которые зарезервированы под хранение GrammarInfo),
+		// начиная счет с нуля.
+		return b == 100;
```
Так 95 или 100?

demidko replied:
100, 95и уже нигде нет.

bazhenov replied:
Но у тебя осталось дублирование с `com.farpost.aot.Compiler.byteFromChar()`

demidko committed:
Исправления

demidko committed:
Заменил абстрактный класс на общий статический метод

demidko committed:
Рефакторинг

demidko committed:
Идея для реализации поиска всех флексий всех лемм,
по одной входной флексии.

Commit:
Update README.md

bazhenov commented on library/src/main/java/com/farpost/aot/data/GrammarInfo.java:
```diff
@@ -0,0 +1,92 @@
+ * Ни в коем случае нельзя менять порядок элементов этого перечисления,
+ * так как от их порядка зависит преобразование байта из файла, в нужный элемент перечисления!
+ */
+public enum GrammarInfo {
```
теги связанные с частью речи (POS) выделить в отдельный enum и сделать getter в `FlexionInfo`

demidko committed:
Обновил ReadMe

demidko committed:
Выделил отдельное перечисление с частями речи.

demidko committed:
Выделил отдельное перечисление с частями речи.

demidko committed:
Новая версия.

demidko committed:
Новая версия.

demidko committed:
benchmark

demidko committed:
benchmark

demidko committed:
оптимизация

demidko committed:
оптимизация

demidko committed:
оптимизация

demidko committed:
оптимизация

Commit:
Merge pull request #2 from demidko/loading-problem

Loading problem

demidko committed:
комментарий

demidko committed:
комментарий

demidko committed:
Merge remote-tracking branch 'origin/master'

demidko committed:
Новая версия

demidko committed:
Новая версия. Устранена избыточная проверка замедлявшая производительность.

bazhenov commented on benchmark/pom.xml:
```diff
@@ -0,0 +1,67 @@
+		</plugins>
+	</build>
+
+	<artifactId>bencmark</artifactId>
```
Давай `artifactId` поднимем наверх (сразу после тега `parent`), а то один из ключевых тегов (имя модуля) в самом конце файла.

demidko replied:
исправил

demidko committed:
Версия 2.7: Увеличена скорость lookup на 1000 ops/ms

bazhenov commented on benchmark/src/main/java/com/farpost/aot/LookupBenchmark.java:
```diff
@@ -0,0 +1,81 @@
+
+		final List<String> words = new ArrayList<>();
+		final AtomicInteger wordId = new AtomicInteger(0);
+		String nextWord;
```
Поля можно сделать `private`

demidko replied:
сделал

bazhenov commented on compiler/pom.xml:
```diff
@@ -0,0 +1,52 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Нарушено форматирование. Проверь настройки IDE.

bazhenov commented on compiler/src/main/java/com/farpost/aot/CollisionsFilter.java:
```diff
@@ -0,0 +1,52 @@
+
+public class CollisionsFilter {
+
+	public static class Result {
```
Даже если у класса все свойства `final` лучше getter'ы все равно выделять и замыкать доступ через них.

demidko replied:
ок готово

bazhenov commented on compiler/src/main/java/com/farpost/aot/CollisionsFilter.java:
```diff
@@ -0,0 +1,52 @@
+	private static boolean isCollision(List<Flexion> flexionsWithOneHash) {
+		return flexionsWithOneHash.stream()
+			.map(x -> x.source)
+			.collect(Collectors.toSet())
```
Старайся использовать все возможности языка, которые позволяют делать код менее многословным. В данном случае, очень уместен static import – `.collect(toSet())`.

bazhenov commented on compiler/src/main/java/com/farpost/aot/CollisionsFilter.java:
```diff
@@ -0,0 +1,52 @@
+		}
+	}
+
+	private static boolean isCollision(List<Flexion> flexionsWithOneHash) {
```
Здесь достаточно `Collection`. В аргументах лучше использовать настолько общие типы, насколько это позволяет контракт.

bazhenov commented on compiler/src/main/java/com/farpost/aot/CollisionsFilter.java:
```diff
@@ -0,0 +1,52 @@
+		var hashToFlex = new HashMap<Integer, List<Flexion>>();
+		for (var i : flexions) {
+			hashToFlex
+				.computeIfAbsent(i.source.hashCode(), k -> new ArrayList<>())
```
Если контрольная сумма флексии настолько важна для словаря, я думаю стоит выделить отдельный метод в `Flexion` для этой задачи.

demidko replied:
Написал тест сравнивающий хеши jvm8 и jvm11 (эквивалентны)

bazhenov commented on compiler/src/main/java/com/farpost/aot/Compiler.java:
```diff
@@ -0,0 +1,74 @@
+final class Compiler {
+	public static void main(String[] args) throws IOException {
+		System.out.println("Парсим /mrd . . .");
+		final var store = new FlexionStorage();
```
Не нужно пихать `final` в каждую локальную переменную. Это затрудняет чтение кода, ставь там где это действительно важно.

demidko replied:
Ок

bazhenov commented on compiler/src/main/java/com/farpost/aot/Compiler.java:
```diff
@@ -0,0 +1,74 @@
+
+
+final class Compiler {
+	public static void main(String[] args) throws IOException {
```
Этот метод читается крайне сложно. Перепиши его, разбив на методы таким образом, чтобы по коду были ясны отдельные фазы компиляции.

demidko replied:
Готово

bazhenov commented on compiler/src/main/java/com/farpost/aot/Compiler.java:
```diff
@@ -0,0 +1,74 @@
+		System.out.println(String.format("Всех флексий %d", normal.size()));
+		System.out.println("Извлекаем коллизии . . .");
+		final var removed = CollisionsFilter.extractCollisions(normal).removedCollisions;
+		System.out.println(String.format("Извлечено колизионных флексий %d", removed.size()));
```
Все комментарии должны быть на английском. Агрессивней используй static import. Много лишних упоминаний классов

bazhenov commented on compiler/src/main/java/com/farpost/aot/Compiler.java:
```diff
@@ -0,0 +1,74 @@
+	private static byte[] bytesFromGrammars(final List<GrammarInfo> line) {
+		final var res = new byte[line.size() + 1];
+		for (var i = 0; i < (res.length - 1); ++i) {
+			res[i] = (byte) Arrays.binarySearch(GrammarInfo.values(), line.get(i));
```
Если ты здесь результат кастишь к byte, то было бы неплохо сначала проверить размер массива.

demidko replied:
Проверял, он меньше чем 90

demidko replied:
Добавил проверку

bazhenov commented on library/pom.xml:
```diff
@@ -0,0 +1,47 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0"
```
Давай явно укажем компилятору, что этот модуль должен быть собран под java 1.8. Иначе результат, насколько я понимаю, может зависеть от того на какой JVM собирается проект.

demidko replied:
в родительском POM прописана компиляция под JVM 8, поэтому дочерний наследует это указание

bazhenov commented on library/src/main/java/com/farpost/aot/LemmaDictionary.java:
```diff
@@ -0,0 +1,70 @@
+		List<LemmaInfo> res = new ArrayList<>();
+		for (int i = 0; i < indexes.length; i += 2) {
+
+			boolean notFound = true;
```
Булевые переменные без отрицания в имени проще поддаются понимаю.

demidko replied:
ок

bazhenov commented on library/src/main/java/com/farpost/aot/data/FlexionInfo.java:
```diff
@@ -0,0 +1,74 @@
+	 *
+	 * @param info набор грамматической информации
+	 */
+	public boolean isNot(final GrammarTag... info) {
```
Более удачное название для метода `notAnyOf`

bazhenov commented on library/src/main/java/com/farpost/aot/data/FlexionInfo.java:
```diff
@@ -0,0 +1,74 @@
+import java.util.Arrays;
+import java.util.List;
+
+public class FlexionInfo {
```
Нужно выделить так же как и `PartOfSpeech` информацию о:

* роде;
* числе;
* падеже;
* одушевленности.

demidko replied:
Невозможно одназначено выделить ни один из этих признаков

bazhenov commented on library/src/main/java/com/farpost/aot/data/GrammarTag.java:
```diff
@@ -0,0 +1,92 @@
+ * Ни в коем случае нельзя менять порядок элементов этого перечисления,
+ * так как от их порядка зависит преобразование байта из файла, в нужный элемент перечисления!
+ */
+public enum GrammarTag {
```
Корректней назвать `MorphologyTag`, так как все эти признаки морфологические, а не грамматические.

bazhenov commented on library/src/main/java/com/farpost/aot/data/LemmaInfo.java:
```diff
@@ -0,0 +1,35 @@
+	// каждая флексия совпадает по строке с искомой
+	public final List<FlexionInfo> flexions = new ArrayList<>();
+	// лемма
+	public final String lemma;
```
getter!

bazhenov commented on library/src/main/java/com/farpost/aot/data/LemmaInfo.java:
```diff
@@ -0,0 +1,35 @@
+
+	public LemmaInfo(final int lemmaIndex, final String lemma) {
+		this.lemmaIndex = lemmaIndex;
+		this.lemma = lemma;
```
`requireNonNull`

bazhenov commented on library/src/main/java/com/farpost/aot/data/PartOfSpeech.java:
```diff
@@ -0,0 +1,87 @@
+package com.farpost.aot.data;
+
+public enum PartOfSpeech {
```
А зачем нам аналогичные теги в `GrammarInfo`? Можем ли мы обойтись одним enum'ом?

demidko replied:
Не можем, тогда поломается декомпиляция, которая сейчас построена на том что каждый байт является порядковым номером соответствующего элемента MorphologyTag. 
У элементов PartOfSpeech другие порядковые номера. 
Если мы захотим разделить эти перечисления на этапе считывания, то усложним код + просядет скорость загрузки.
Если захотим разделить их на этапе компиляции, то и усложним компиляцию и скорость загрузки опять таки просядет.

bazhenov commented on library/src/main/java/com/farpost/aot/func/Decompiler.java:
```diff
@@ -0,0 +1,52 @@
+ */
+public final class Decompiler {
+
+	// оптимизация чтобы не выделять каждый раз память под буфер юникодных символов
```
Комментарии надо писать через
```
/**
 *
 */
```

Тогда они корректно показываются средой по Ctrl+J

bazhenov commented on library/src/main/java/com/farpost/aot/func/Decompiler.java:
```diff
@@ -0,0 +1,52 @@
+
+	// оптимизация чтобы не выделять каждый раз память под буфер юникодных символов
+	// 36 - длина наибольшей флексии, поэтому больше нам и не потребуется
+	private static final char[] resultBuffer = new char[36];
```
Что произойдет с этим кодом если словарь будет использоваться в многопоточной среде?

bazhenov commented on library/src/main/java/com/farpost/aot/storages/CollisionFlexionStorage.java:
```diff
@@ -0,0 +1,59 @@
+			final int lemmaIndex = reader.readInt();
+			final int grammarIndex = reader.readInt();
+
+			final byte[] strbuf = new byte[36];
```
"Магические числа" должны быть выделены в константы с описанием

bazhenov commented on library/src/main/java/com/farpost/aot/storages/CollisionFlexionStorage.java:
```diff
@@ -0,0 +1,59 @@
+				strbuf[++strbufIndex] = j;
+			}
+
+			final String javaString = stringFromBytes(Arrays.copyOf(strbuf, strbufIndex + 1));
```
Вроде, тут можно без копирования обойтись

demidko replied:
Исправил

bazhenov commented on library/src/main/java/com/farpost/aot/storages/GrammarStorage.java:
```diff
@@ -0,0 +1,47 @@
+			int bufIndex = -1;
+
+			// считываем строку
+			for (byte currentByte = reader.readByte(); !isEndl(currentByte); currentByte = reader.readByte()) {
```
Дублирование. Аналогичный код в `CollisionFlexionStorage`

bazhenov commented on library/src/main/java/com/farpost/aot/storages/LemmaStorage.java:
```diff
@@ -0,0 +1,51 @@
+	private final String[] lines;
+
+	public LemmaStorage(DataInputStream reader) throws IOException {
+		/*lines = new byte[reader.readInt()][];
```
Мертвый код нужно удалять

bazhenov commented on library/src/main/java/com/farpost/aot/storages/NormalFlexionStorage.java:
```diff
@@ -0,0 +1,71 @@
+	}
+
+
+	private static int intFromBytes(byte a, byte b, byte c, byte d) {
```
Судя по использованию, какой-то неудобный helper. Можешь придумать более подходящий контракт исходя из контекста использования метода?

demidko replied:
ок

bazhenov commented on library/src/test/java/com/farpost/aot/LemmaDictionaryTest.java:
```diff
@@ -0,0 +1,82 @@
+
+	@Test
+	public void dictionaryShouldBeAbleToFindExistentWords() throws IOException {
+		assertThat(d.lookup("краснеющий").size(), is(1));
```
Есть более подходящий matcher `hasSize()`

bazhenov commented on library/src/test/java/com/farpost/aot/LemmaDictionaryTest.java:
```diff
@@ -0,0 +1,82 @@
+
+	private final LemmaDictionary d = new LemmaDictionary();
+
+	public LemmaDictionaryTest() throws IOException {
```
Бесполезный конструктор

demidko replied:
Он нужен т. к. 
private final LemmaDictionary d = new LemmaDictionary();
Выкидывает прописанный в конструкторе эксцепшн

bazhenov commented on loader/pom.xml:
```diff
@@ -0,0 +1,41 @@
+			<version>6.14.3</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
```
тестовые зависимости полезно определять один раз в родительском проекте

bazhenov commented on loader/src/main/java/com/farpost/aot/FlexionStorage.java:
```diff
@@ -0,0 +1,70 @@
+			// !!! вывод всех неоднозначностей
+			if (flex.size() > 2 && flex.get(0).source.equals(flex.get(1).source)) {
+				System.out.println("Неоднозначность! База '" + toks[0].toLowerCase().replace('ё', 'е') + "', парадигма склонения:");
+				for (final var j : flex) {
```
То же самое можно сделать более "идиоматично":

```
String lemmas = flex.stream()
  .map(Flexion::getSource)
  .collect(joining(" "));
out.println(lemmas);
```

bazhenov commented on loader/src/main/java/com/farpost/aot/FlexionStorage.java:
```diff
@@ -0,0 +1,70 @@
+				System.out.println("Неоднозначность! База '" + toks[0].toLowerCase().replace('ё', 'е') + "', парадигма склонения:");
+				for (final var j : flex) {
+					System.out.print(j.source);
+					System.out.print(' ');
```
Если ты часто используешь `System.out`, то можно сделать static import.

bazhenov commented on loader/src/main/java/com/farpost/aot/func/Mod.java:
```diff
@@ -0,0 +1,28 @@
+		this.post = post;
+	}
+
+	public Flexion apply(final String base) {
```
Очень похоже на функциональное API. Может стоит ли имплементировать `Function`?

bazhenov commented on loader/src/main/java/com/farpost/aot/readers/TabReader.java:
```diff
@@ -0,0 +1,23 @@
+
+import static com.farpost.aot.readers.BufferedReaderFabric.createUtf8Reader;
+
+public class TabReader {
```
Не очень удачное, на мой взгляд, разделение ответственностей. Единственный клиент `TabReader` – `GrammarStorage`. И он потом начинает парсить содержимое `tab` файла из строки. Почему бы это не сделать здесь?

bazhenov commented on loader/src/main/java/com/farpost/aot/storages/ParadigmStorage.java:
```diff
@@ -0,0 +1,41 @@
+						final var infoIndex = gram.getVariantIndex(args[1]);
+						return args.length == 2 ?
+							new Mod(infoIndex, null, postfix) :
+							new Mod(infoIndex, args[2].toLowerCase().replace('ё', 'е'), postfix);
```
Дублирование обработки буквы ё

bazhenov commented on loader/src/main/java/com/farpost/aot/storages/ParadigmStorage.java:
```diff
@@ -0,0 +1,41 @@
+				Arrays.stream(reader.readLine().split("%"))
+					.filter(s -> !s.isBlank())
+					.map(src -> {
+						final var args = src.split("\\*");
```
Давай выделим это замыкание в отдельный метод. Слишком сложный pipeline уже получается

bazhenov commented on testapp/src/main/java/com/farpost/aot/TestApplication.java:
```diff
@@ -0,0 +1,29 @@
+		try (final BufferedReader reader = new BufferedReader(
+			new InputStreamReader(System.in)
+		)) {
+			while (true) {
```
Это не важно в данном случае, поэтому исправлять не обязательно, но может быть тебе будет полезно знать.

Лучше вечные циклы заменять на:

```
while (currentThread().isInterrupted()) {

}
```

Это позволяет прерывать такие циклы извне. В противном случае может оказаться так (зависит от того как создавался поток), что единственный способ остановить приложение - `kill -9`.

demidko committed:
Рефакторинг.

demidko committed:
Рефакторинг.

demidko committed:
Рефакторинг.

demidko committed:
Рефакторинг.

demidko committed:
Улучшения

demidko committed:
Улучшения ещё

demidko committed:
Исправления для код ревью

demidko committed:
Убрал некорректный метод леммы getPartOfSpeech,т. к. флексии могут несовпадать с леммой по части речи.

demidko committed:
Убрал некорректный метод леммы getPartOfSpeech,т. к. флексии могут не совпадать с леммой по части речи.

demidko committed:
Исправил бесконечный цикл

demidko committed:
Исправил бесконечный цикл

demidko committed:
Начата разработка нового компилятора (для решения задачи полного поиска по флексии).

demidko committed:
Начата разработка нового компилятора (для решения задачи полного поиска по флексии).

Commit:
Update README.md

Commit:
Update README.md

demidko committed:
Ускоряю компиляцию

demidko committed:
Merge remote-tracking branch 'origin/master'

demidko committed:
Ускоряю компиляцию2

demidko committed:
рефакторинг

demidko committed:
refactoring

demidko committed:
исправление

bazhenov commented on library/src/main/java/com/farpost/aot/LemmaDictionary.java:
```diff
@@ -0,0 +1,70 @@
+public class LemmaDictionary {
+
+	private final CollisionFlexionStorage colFlex;
+	private final NormalFlexionStorage norFlex;
```
Предлагаю перенести `CollisionFlexionStorage` и `NormalFlexionStorage` в пакет `com.farpost.aot`. Это позволит сделать эти классы package-local. Им совсем не обязательно быть публичными.

demidko replied:
Готово

bazhenov commented on library/src/main/java/com/farpost/aot/data/LemmaInfo.java:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	@Override
+	public String toString() {
```
```
String flexions = this.flexions.stream()
    .map(Object::toString)
    .collect(joining(", "));
return String.format("<%s %s>", lemma, flexions);
```

Согласись проще читается

bazhenov commented on library/src/main/java/com/farpost/aot/data/FlexionInfo.java:
```diff
@@ -0,0 +1,78 @@
+	/**
+	 * Возвращает иммутабельную коллекцию тегов
+	 *
+	 * @return список тегов
```
В таких случаях вполне уместно совмещать основной блок описания и тег `@return`. Т.е.

```
/**
 * @return множество морфологических тегов
 */
```

demidko replied:
Ок, готово

bazhenov commented on library/src/main/java/com/farpost/aot/data/FlexionInfo.java:
```diff
@@ -0,0 +1,78 @@
+	 *
+	 * @return список тегов
+	 */
+	public List<MorphologyTag> getAllTags() {
```
Может уместней сделать возвращаемое значение `Set<MorphologyTag>`?

bazhenov commented on library/src/main/java/com/farpost/aot/func/Decompiler.java:
```diff
@@ -0,0 +1,70 @@
+	 * Читает из потока байт юникодную строку
+	 * @param reader поток байт
+	 * @return строка
+	 * @throws IOException
```
Пустые теги никому не нужны, они не добавляют информации и их не следует писать.

bazhenov commented on library/src/main/java/com/farpost/aot/func/Decompiler.java:
```diff
@@ -0,0 +1,70 @@
+	/**
+	 * Аналогичная оптимизация для грамматики
+	 */
+	private static final MorphologyTag[] grmBuf = new MorphologyTag[12];
```
Поля выше методов

demidko committed:
Исправления стиля

demidko committed:
Теперь компиляция ресурсов при сборке работает корректно.

demidko committed:
3.3

demidko committed:
Форматирование

demidko committed:
Исправление компилятора

demidko committed:
Файл для сохранения папки /resources в git

demidko committed:
Исправил баг коллизий с не-флексионными строками.

demidko committed:
Сохранение
